pkg/config/database.go

```diff
@@ -75,7 +81,46 @@ func (d *Database) Open(logger *logging.Logger) (*icingadb.DB, error) {
 			}
 		}
 
-		dsn = config.FormatDSN()
+		c, err := mysql.NewConnector(config)
+		if err != nil {
+			return nil, errors.Wrap(err, "can't open mysql database")
+		}
+
+		wsrepSyncWait := int64(d.Options.WsrepSyncWait)
+
+		// Setting "wsrep_sync_wait" for each session ensures that causality checks are performed before execution
+		// and that each statement is executed on a fully synchronized node. Doing so prevents foreign key violation
+		// when inserting into dependent tables on different MariaDB/MySQL nodes. When using MySQL single nodes, the
+		// "SET SESSION" command will fail with "Unknown system variable (1193)" and will therefore be silently dropped.
+		// https://mariadb.com/kb/en/galera-cluster-system-variables/#wsrep_sync_wait
+		var setGaleraOpts = func(ctx context.Context, conn driver.Conn) error {
+			const galeraOpts = "SET SESSION wsrep_sync_wait=?"
+
+			stmt, err := conn.(driver.ConnPrepareContext).PrepareContext(ctx, galeraOpts)
+			if err != nil {
```

Comment:
That whole error handling looks overdone. The prepare returns `errUnknownSysVar` not the exec.